pkg/server/apiextensions.go

```diff
@@ -245,117 +259,226 @@ func (c *apiBindingAwareCRDLister) ListWithContext(ctx context.Context, selector
-		}
+	if err != nil {
+		return nil, err
+	}
 
-		return crd, nil
+	if isPartialMetadataRequest(ctx) {
+		crd = partialMetadataCRD(crd)
 	}
 
-	// Priority 1: see in system CRDs
-	systemCRDKeys := c.systemCRDProvider.Keys(ctx, cluster.Name)
-	systemCRDKeyName := clusters.ToClusterAwareKey(SystemCRDLogicalCluster, name)
-	if systemCRDKeys.Has(systemCRDKeyName) {
-		crd, err = c.crdLister.Get(systemCRDKeyName)
-		if err != nil && !apierrors.IsNotFound(err) {
-			// something went wrong w/the lister - could only happen if meta.Accessor() fails on an item in the store.
-			return nil, err
+	return crd, nil
+}
+
+func partialMetadataCRD(crd *apiextensionsv1.CustomResourceDefinition) *apiextensionsv1.CustomResourceDefinition {
+	crd = crd.DeepCopy()
+
+	if crd.Annotations == nil {
+		crd.Annotations = make(map[string]string)
+	}
+
+	crd.Annotations[annotationKeyPartialMetadata] = ""
+
+	// set minimal schema that prunes everything but ObjectMeta
+	for _, v := range crd.Spec.Versions {
+		v.Schema = &apiextensionsv1.CustomResourceValidation{
+			OpenAPIV3Schema: &apiextensionsv1.JSONSchemaProps{
+				Type: "object",
+			},
 		}
+	}
+
+	return crd
+}
+
+func (c *APIBindingAwareCRDLister) getForFullData(clusterName logicalcluster.LogicalCluster, name string) (*apiextensionsv1.CustomResourceDefinition, error) {
+	if clusterName == logicalcluster.Wildcard {
+		return c.getWildcard(name)
+	}
+
+	return c.get(clusterName, name)
+}
+
+func (c *APIBindingAwareCRDLister) getWildcard(name string) (*apiextensionsv1.CustomResourceDefinition, error) {
+	// TODO use an index by group+resource
+	crds, err := c.crdLister.List(labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+
+	crd, equivalentSchemas := findCRD(name, crds)
+	if !equivalentSchemas {
+		err = apierrors.NewInternalError(fmt.Errorf("error resolving resource: cannot watch across logical clusters for a resource type with several distinct schemas"))
+		return nil, err
+	}
+
+	if crd == nil {
+		return nil, errors.NewNotFound(apiextensionsv1.Resource("customresourcedefinitions"), name)
+	}
+
+	return crd, nil
+}
+
+func (c *APIBindingAwareCRDLister) getForIdentity(name, identity string) (*apiextensionsv1.CustomResourceDefinition, error) {
+	apiExports, err := c.apiExportIndexer.ByIndex("byIdentity", identity)
+	if err != nil {
+		return nil, err
+	}
 
-		if crd != nil {
-			return crd, nil
+	if len(apiExports) == 0 {
+		return nil, errors.NewNotFound(apiextensionsv1.Resource("customresourcedefinitions"), name)
+	}
+
+	group, resource := crdNameToGroupResource(name)
+
+	// If for some reason multiple APIExports share the same identity, that shouldn't matter. We just need any
+	// one of them.
+	apiExport := apiExports[0].(*apisv1alpha1.APIExport)
+	apiExportClusterName := logicalcluster.From(apiExport)
+
+	var boundCRDName string
+
+	for _, schemaName := range apiExport.Spec.LatestResourceSchemas {
```

Comment:
the export schemas can be totally different by now. The binding should count, until transactionally we switch over to new schemas and store them in the binding status.